File: src/pocketcloud/cloud/plugin/CloudPluginManager.php
<?php

namespace pocketcloud\cloud\plugin;

use pocketcloud\cloud\event\EventManager;
use pocketcloud\cloud\event\impl\plugin\PluginDisableEvent;
use pocketcloud\cloud\event\impl\plugin\PluginEnableEvent;
use pocketcloud\cloud\event\impl\plugin\PluginLoadEvent;
use pocketcloud\cloud\plugin\loader\CloudPluginLoader;
use pocketcloud\cloud\plugin\loader\FolderCloudPluginLoader;
use pocketcloud\cloud\plugin\loader\PharCloudPluginLoader;
use pocketcloud\cloud\terminal\log\CloudLogger;
use pocketcloud\cloud\util\SingletonTrait;
use pocketcloud\cloud\util\tick\Tickable;
use Throwable;

final class CloudPluginManager implements Tickable {
    use SingletonTrait;

    /** @var array<CloudPlugin> */
    private array $plugins = [];
    /** @var array<CloudPlugin> */
    private array $enabledPlugins = [];
    /** @var array<CloudPluginLoader> */
    private array $loaders = [];

    public function __construct() {
        self::setInstance($this);
        $this->registerLoader(new PharCloudPluginLoader());
        $this->registerLoader(new FolderCloudPluginLoader());
    }

    public function registerLoader(CloudPluginLoader $loader): void {
        $this->loaders[] = $loader;
    }

    public function loadAll(): void {
        CloudLogger::get()->debug("Loading plugins...");
        foreach (array_diff(scandir(CLOUD_PLUGINS_PATH), [".", ".."]) as $file) {
            $path = CLOUD_PLUGINS_PATH . $file;
            $this->load($path);
        }

        if (count($this->plugins) == 0) {
            CloudLogger::get()->info("No plugins were loaded.");
        } else {
            CloudLogger::get()->success("Successfully loaded §b" . count($this->plugins) . " plugin" . (count($this->plugins) == 1 ? "" : "s") . "§r.");
        }
    }

    public function load(string $path): void {
        CloudLogger::get()->debug("Loading plugim §b" . basename($path) . "§r...");
        foreach ($this->loaders as $loader) {
            try {
                if ($loader->canLoad($path)) {
                    $plugin = $loader->loadPlugin($path);
                    if (!$plugin instanceof CloudPlugin) {
                        CloudLogger::get()->error("§cFailed to load the plugin §e" . basename($path) . "§c: §eMain Class does not inherit from §eCloudPlugin");
                        return;
                    }

                    if (isset($this->plugins[$plugin->getDescription()->getName()])) {
                        CloudLogger::get()->warn("§cThe plugin §e" . $plugin->getDescription()->getName() . " §cis already loaded.");
                        return;
                    }

                    new PluginLoadEvent($plugin)->call();
                    $this->plugins[$plugin->getDescription()->getName()] = $plugin;
                    $plugin->onLoad();
                }
            } catch (Throwable $exception) {
                CloudLogger::get()->error("§cFailed to load the plugin §e" . basename($path) . "§c: §e" . $exception->getMessage());
                CloudLogger::get()->exception($exception);
            }
        }
    }

    public function enableAll(): void {
        CloudLogger::get()->debug("Enabling plugins...");
        foreach ($this->plugins as $plugin) {
            $this->enable($plugin);
        }

        if (count($this->enabledPlugins) > 0) {
            CloudLogger::get()->success("Successfully enabled §b" . count($this->enabledPlugins) . " plugin" . (count($this->enabledPlugins) == 1 ? "" : "s") . "§r.");
        }
    }

    public function enable(CloudPlugin $plugin): void {
        CloudLogger::get()->info("Enabling §b" . $plugin->getDescription()->getName() . "§r...");
        $plugin->setEnabled(true);
        new PluginEnableEvent($plugin)->call();
        try {
            $plugin->onEnable();
        } catch (Throwable $throwable) {
            CloudLogger::get()->exception($throwable);
            $this->disable($plugin);
        }

        if ($plugin->isEnabled()) {
            $this->enabledPlugins[$plugin->getDescription()->getName()] = $plugin;
        }
    }

    public function disableAll(): void {
        foreach ($this->enabledPlugins as $plugin) {
            $this->disable($plugin);
        }

        CloudLogger::get()->info("Disabled all plugins.");
    }

    public function disable(CloudPlugin $plugin): void {
        CloudLogger::get()->info("Disabling §b" . $plugin->getDescription()->getName() . "§r...");
        new PluginDisableEvent($plugin)->call();
        $plugin->setEnabled(false);
        $plugin->onDisable();

        $plugin->getScheduler()->cancelAll();
        EventManager::getInstance()->removeHandlers($plugin);
        if (isset($this->enabledPlugins[$plugin->getDescription()->getName()])) unset($this->enabledPlugins[$plugin->getDescription()->getName()]);
    }

    public function clear(): void {
        $this->plugins = [];
        $this->enabledPlugins = [];
    }

    public function tick(int $currentTick): void {
        foreach ($this->enabledPlugins as $enabledPlugin) {
            if ($enabledPlugin->isEnabled()) {
                $enabledPlugin->getScheduler()->tick($currentTick);
            }
        }
    }

    public function get(string $name): ?CloudPlugin {
        return $this->plugins[$name] ?? null;
    }

    public function getLoaders(): array {
        return $this->loaders;
    }

    public function getAll(bool $enabled = false): array {
        return $enabled ? $this->enabledPlugins : $this->plugins;
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/plugin/CloudPluginManager.java
package de.pocketcloud.cloud.plugin;

import de.pocketcloud.cloud.PocketCloud;
import de.pocketcloud.cloud.console.log.CloudLogger;
import de.pocketcloud.cloud.event.impl.plugin.PluginDisableEvent;
import de.pocketcloud.cloud.event.impl.plugin.PluginEnableEvent;
import de.pocketcloud.cloud.event.impl.plugin.PluginLoadEvent;
import de.pocketcloud.cloud.plugin.exception.PluginLoadFailedException;
import de.pocketcloud.cloud.plugin.loader.JarCloudPluginLoader;
import de.pocketcloud.common.lifecycle.Loadable;
import de.pocketcloud.common.lifecycle.Tickable;
import lombok.Getter;

import java.io.IOException;
import java.nio.file.DirectoryStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

@Getter
public final class CloudPluginManager implements Tickable, Loadable {

    private final Map<String, CloudPlugin> plugins = new ConcurrentHashMap<>();
    private final JarCloudPluginLoader pluginLoader = new JarCloudPluginLoader();
    private final Path pluginsFolder = Paths.get("storage/plugins");

    public void load() {
        CloudLogger.get().info("Loading plugins...");
        List<CloudPlugin> eligiblePlugins = new ArrayList<>();
        try (DirectoryStream<Path> stream = Files.newDirectoryStream(pluginsFolder)) {
            for (Path path : stream) {
                if (Files.isRegularFile(path)) {
                    if (path.getFileName().toString().endsWith(".jar")) {
                        CloudPlugin pl = loadPlugin(path);
                        if (pl != null) eligiblePlugins.add(pl);
                    }
                }
            }
        } catch (IOException e) {
            CloudLogger.get().exception("Unable to load plugins", e);
        } finally {
            for (CloudPlugin pl : eligiblePlugins) {
                enable(pl);
            }
        }
    }

    @Override
    public void unload() {
        disableAll();
        this.plugins.clear();
    }

    public CloudPlugin loadPlugin(Path jarFile) {
        if (pluginLoader.canLoad(jarFile)) {
            CloudLogger.get().info("Loading plugin §b{}§r...", jarFile.getFileName().toString());
            CloudPlugin plugin;

            try {
                plugin = pluginLoader.load(jarFile);
            } catch (Exception e) {
                CloudLogger.get().exception("Unable to load plugin " + jarFile.getFileName(), e);
                return null;
            }

            try {
                if (plugins.containsKey(plugin.getDescription().name())) throw new PluginLoadFailedException("Plugin with the same name already loaded");
                plugins.put(plugin.getDescription().name(), plugin);
                new PluginLoadEvent(plugin).call();
                plugin.onLoad();
                return plugin;
            } catch (Exception e) {
                CloudLogger.get().exception("Unable to load plugin " + jarFile.getFileName(), e);
                disable(plugin);
            }
        }

        return null;
    }

    public void enableAll() {
        plugins.values().stream().filter(CloudPlugin::isDisabled).forEach(this::enable);
    }

    public void enable(CloudPlugin plugin) {
        if (plugin.isEnabled()) return;
        CloudLogger.get().info("§aEnabling §rplugin §b{}§r...", plugin.getDescription().name());
        plugin.setState(CloudPluginState.ENABLED);
        try {
            new PluginEnableEvent(plugin).call();
            plugin.onEnable();
        } catch (Exception e) {
            plugin.getLogger().exception("Unable to enable plugin", e);
            disable(plugin);
        }
    }

    public void disableAll() {
        CloudLogger.get().info("§cDisabling §rall plugins...");
        plugins.values().stream().filter(CloudPlugin::isEnabled).forEach(this::disable);
    }

    public void disable(CloudPlugin plugin) {
        if (plugin.isDisabled()) return;
        CloudLogger.get().info("§cDisabling §rplugin §b{}§r...", plugin.getDescription().name());
        plugins.remove(plugin.getDescription().name());
        plugin.setState(CloudPluginState.DISABLED);
        PocketCloud.instance().events().unregisterAll(plugin);
        try {
            new PluginDisableEvent(plugin).call();
            plugin.onDisable();
        } catch (Exception e) {
            plugin.getLogger().exception("Unable to disable plugin", e);
        } finally {
            try {
                plugin.getLoader().close();
            } catch (Exception _) {}
        }
    }

    @Override
    public void tick(long currentTick) {
        for (CloudPlugin plugin : plugins.values()) {
            if (!plugin.isEnabled()) continue;
            plugin.getScheduler().tick(currentTick);
        }
    }

    public int pluginCount() {
        return plugins.size();
    }

    public Optional<CloudPlugin> get(String name) {
        return Optional.ofNullable(plugins.get(name));
    }

    public List<CloudPlugin> getEnabledPlugins() {
        return plugins.values().stream().filter(CloudPlugin::isEnabled).collect(Collectors.toList());
    }

    public List<CloudPlugin> getDisabledPlugins() {
        return plugins.values().stream().filter(CloudPlugin::isDisabled).collect(Collectors.toList());
    }
}